src/librustc_resolve/lib.rs

```diff
@@ -2700,18 +2702,27 @@ impl<'a> Resolver<'a> {
                         }
                         return (err, candidates);
                     },
-                    _ if ns == ValueNS && is_struct_like(def) => {
-                        if let Def::Struct(def_id) = def {
-                            if let Some((ctor_def, ctor_vis))
-                                    = this.struct_constructors.get(&def_id).cloned() {
-                                if is_expected(ctor_def) && !this.is_accessible(ctor_vis) {
-                                    err.span_label(span, format!("constructor is not visible \
-                                                                   here due to private fields"));
-                                }
+                    (Def::Struct(def_id), _) if ns == ValueNS => {
```

Comment:
`Def::Union(..)`, `Def::Variant(..)` and `Def::VariantCtor(_, CtorKind::Fictive)` can be added here to restore what the previous code did (they are equivalent to structs without accessible constructors in this context).

Reply:
Only `CtorKind::Fictive` is relevant here because for others https://github.com/rust-lang/rust/pull/47116#discussion_r161618498 still holds.